code/controllers/subsystem/html_audio.dm

```diff
@@ -0,0 +1,274 @@
+
+				if(hearer_env && source_env)
+					var/pressure = min(hearer_env.return_pressure(), source_env.return_pressure())
+					if(pressure < ONE_ATMOSPHERE)
+						pressure_factor = max((pressure - SOUND_MINIMUM_PRESSURE)/(ONE_ATMOSPHERE - SOUND_MINIMUM_PRESSURE), 0)
+				else //space
+					pressure_factor = 0
+
+				if(distance <= 1)
+					pressure_factor = max(pressure_factor, 0.15) //touching the source of the sound
+
+				volume_to_use *= pressure_factor
+				volume_to_use = clamp(volume_to_use, 0, 1)
+		listener << output(list2params(list(num2text(volume_to_use), "channel_[i]")), "html_audio_player:setVolume")
+
+/datum/controller/subsystem/html_audio/proc/jank_ass_browse_check(checked_person)
+	if (!winexists(checked_person, "html_audio_player"))
+		checked_person << browse(browse_txt, "window=html_audio_player&file=html_audio_player.htm")
+	if (!winexists(checked_person, "html_audio_preview_player"))
+		checked_person << browse(browse_txt, "window=html_audio_preview_player&file=html_audio_preview_player.htm")
+
+/datum/controller/subsystem/html_audio/proc/register_player(atom/movable/player, requires_LOS = FALSE)
+	var/assigned = FALSE
+	for(var/i in 1 to max_channels) // Go through the channels, find the first empty slot and get in there.
+		if(isnull(channel_assignment[i]))
+			assigned = TRUE
+			channel_assignment[i] = player
+			if(requires_LOS)
+				channel_requires_LOS_at_start[i] = TRUE
+			RegisterSignal(player, COMSIG_MOVABLE_MOVED, PROC_REF(handle_player_move))
+			RegisterSignal(player, COMSIG_PARENT_QDELETING, PROC_REF(deregister_player_qdel))
+			break
+	if(!assigned)
+		CRASH("YO WE OUT OF FUCKING CHANNELS ABORT ABORT ABORT TOO MANY FUCKING THINGS GOT HTML AUDIO REEL THAT SHIT IN BRO BACK THAT ASS UP")
+	list_clear_nulls(listeners) // clients be like *poof* mid proc
+	for(var/client/listener in listeners)
+		if(!listener)
+			continue
+		update_listener_volume(listener)
+
+/datum/controller/subsystem/html_audio/proc/deregister_player_qdel(atom/movable/player)
+	SIGNAL_HANDLER
+	UnregisterSignal(player, list(COMSIG_MOVABLE_MOVED, COMSIG_PARENT_QDELETING))
+	deregister_player(player)
+
+/datum/controller/subsystem/html_audio/proc/handle_player_move(atom/movable/player, atom/old_loc)
+	SIGNAL_HANDLER
+	list_clear_nulls(listeners) // clients be like *poof* mid proc
+	for(var/client/listener in listeners)
+		if(!listener)
+			continue
+		update_listener_volume(listener)
+
+/datum/controller/subsystem/html_audio/proc/deregister_player(atom/movable/player)
+	UnregisterSignal(player, list(COMSIG_MOVABLE_MOVED, COMSIG_PARENT_QDELETING))
+	var/current_channel = channel_assignment.Find(player)
+	channel_assignment[current_channel] = null
+	active_urls[current_channel] = null
+	channel_loop_status[current_channel] = null
+	channel_tts_status[current_channel] = null
+	channel_requires_LOS_at_start[current_channel] = null
+	channel_requires_LOS_at_start_listeners[current_channel] = null
```

Comment:
Not sure if it should be the case here but do consider `-=`

Reply:
it shouldn't, these are statically sized null arrays but goof plans to redesign this section because looping over 512 channels when only 5 are active is silly.